### DISCUSSION TITLE
Development shutter (USE_SHUTTER_ESP32) and tilt precision

### DISCUSSION THREAD

**Discussion by JosefRypacek** [arendst/Tasmota#18326] [Ideas]
Hi @stefanbode,

I've just compiled dev branch with _USE_SHUTTER_ESP32_ to test your awesome changes as I was trying to setup my first Tasmota device according to the documentation and found that the doc is "more up to date" than code itself. :)

It seems to work for me (Sonoff dual R3 and Venetian Blind) - positioning, tilt, buttons...
The only thing I found is that tilting is probably does not take _ShutterMotorDelay1_ into account.

My blinds have tilt approx 0-90° and 1.05s is exact time to open/close in one step.
`ShutterTiltConfig1 0 90 21 90 0`

When I tried to open in in multiple steps (e.g. 0, 17, 34, 51, 72, 90) a ended with real tilt about 60° instead of 90°.

I know the tilt will never be exact and precise, but I think this can be improved a little bit.

I'm available to test it if needed.

**Comment by stefanbode** [comment]
Thanks for raising this issue. Indeed the tilt combined with the motordelay was never tested. I will check on this

**Reply by JosefRypacek** [reply]
Thanks! I did the calculation for my case and it seems that result (s = 0.08 = 80ms) almost match my motordelay 0.1s I've set for shutter position according to the google spreadsheet. So I think it might work. :)

https://www.wolframalpha.com/input?i=1S+%2B+90D+%3D+1.05%2C+5S+%2B+60D+%3D+1.05
Computed as:
1 Step with duration of 1.05 seconds will result in 90 Degree
5 Steps with overall duration of 1.05 seconds will result in 60 Degree

**Comment by stefanbode** [comment]
ok, I found a solution to get this into the code. No big deal. Will submit later this evening. There is one thing you should keep in mind. The minimum timeslice in TASMOTA is 0.05 seconds. nothing can be shorter. Therefore you motordelay will be 0.05 or 0.1 seconds. You can check what fits better. It will never 100% just some increase in accuracy

**Reply by stefanbode** [reply]
it is merged into the development branch just now

**Reply by JosefRypacek** [reply]
Thanks, will test it.

**Comment by JosefRypacek** [comment]
I have one more question for you. My first device was flashed with FW 12.4.0 and added to Home Assistant as cover (shutters). It remains even after upgrade to dev version 12.4.0.5.
But I've just flashed several additional devices with dev FW 12.4.0.5 and they are showing like switches (relays). I was able to fix it by flashing back to 12.4.0, configuration reset and new configuration.

Is it possible that it may be related to your changes?
I've not found where the problem is yet.

**Comment by stefanbode** [comment]
Where did you get your binary from? The missing shutter buttons in the webinterface was a bug of the first version with 
#define USE_SHUTTER_ESP32
This should be fixed now. If you wan to have a build with the fix you have to compile the current development branch on your own or wait until midnight when develoipment builds gets rebuild. 12.4.0 did not have the ESP32_Shutter and current build only have it if you compile your own version. Anyhow the tilt fix is in the ESP8266 and the ESP32 version

**Reply by JosefRypacek** [reply]
I've compiled own binary with #define USE_SHUTTER_ESP32 from dev yesterday (don't know how to get the commit hash). Web interface is OK, the problem is only in HA - probably MQTT. I'll compare it with stable branch where it works fine and create a proper bug then.

I'll also report how tilting works here.

**Reply by stefanbode** [reply]
Yes please create an "issue" in github. This makes it easiest to follow up separately and fix it. Thanks for your help. I'm not aware to change anything in MQTT. But you never know... If you let me know how HA would like to send position/tilsposition in one command pls. let me know. Current implementation is `shutterposition1 30,-90` or if you want to do it without setting tilt `shutterposition1 30`

**Reply by Mark-Muc** [reply]
i have the same problem .. no real shuttersupport on Shelly Plus 2PM PCB v0.1.9 (button) (no shutterslider)!!
I compiled the dev. brache with:
```
#define USE_SHUTTER_ESP32
#ifndef USE_SHUTTER
#define USE_SHUTTER     
#endif
```
reflshed to 12.4.0 (no #define USE_SHUTTER_ESP32) show the slider again!!

same version on a Sonoff SwitchMan 2C
work like expected!
Strange!!
Mark_Muc

**Comment by JosefRypacek** [comment]
Just compiled `development#b287d0509875347ec200440de405d99de82d60f6` and can confirm that motordelay for tilting is working correctly. I've lowered `ShutterTiltConfig1 0 90 21 90 0` to `ShutterTiltConfig1 0 90 19 90 0` while having `ShutterMotorDelay1 0.1`.

**Comment by JosefRypacek** [comment]
I have Cover control in HA with two sliders (position, tilt), so I can set those parameters separately.
When I change the position, tasmota will restore previous tilt afterwards.
I'm using blinds usually only in a few positions (open, closed with 0°, closed with `45°, closed with 90°), so I will try to create some automation in HA later. 

Just FYI, HA is using following code
https://github.com/home-assistant/core/tree/dev/homeassistant/components/tasmota
which is using following library
https://github.com/emontnemery/hatasmota

```
cmnd/tasmota_48CEEC/ShutterTilt1 54
stat/tasmota_48CEEC/RESULT {"POWER2":"ON"}
stat/tasmota_48CEEC/POWER2 ON
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":38,"Direction":-1,"Target":38,"Tilt":85}}
stat/tasmota_48CEEC/RESULT {"POWER2":"OFF"}
stat/tasmota_48CEEC/POWER2 OFF
stat/tasmota_48CEEC/SHUTTER1 38
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":38,"Direction":0,"Target":38,"Tilt":48}}
```


```
cmnd/tasmota_48CEEC/ShutterPosition1 56
stat/tasmota_48CEEC/RESULT {"POWER1":"ON"}
stat/tasmota_48CEEC/POWER1 ON
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":38,"Direction":1,"Target":56,"Tilt":48}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":38,"Direction":1,"Target":56,"Tilt":48}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":39,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":42,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":45,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":48,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":50,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":53,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":56,"Direction":1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"POWER1":"OFF"}
stat/tasmota_48CEEC/POWER1 OFF
stat/tasmota_48CEEC/RESULT {"POWER2":"ON"}
stat/tasmota_48CEEC/POWER2 ON
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":56,"Direction":-1,"Target":56,"Tilt":90}}
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":56,"Direction":-1,"Target":56,"Tilt":72}}
stat/tasmota_48CEEC/RESULT {"POWER2":"OFF"}
stat/tasmota_48CEEC/POWER2 OFF
stat/tasmota_48CEEC/SHUTTER1 56
stat/tasmota_48CEEC/RESULT {"Shutter1":{"Position":56,"Direction":0,"Target":56,"Tilt":48}}
```

**Reply by JosefRypacek** [reply]
Just FYI, I've created and issue for hatasmota to support position with tilt in single command https://github.com/emontnemery/hatasmota/issues/217. It would be very useful for automation.

I'm currently using manual MQTT Publish in HA to automate my blinds. This new feature you added is must-have for automation. 👍

**Comment by JosefRypacek** [comment]
One more thing... 

My configuration was:
`ShutterTiltConfig1 0 90 19 90 0`
`ShutterButton1 2 0 0/45 0/90`

I tried to reconfigure my device from tilt 0-90° to -90-0°.
I set `ShutterTiltConfig1 -90 0 19 0 -90`, but forgot to update ShutterButton.
When I double pressed the wall switch blinds closed and opened after some time. It gets into weird state and I cannot acces web UI of some devices. Other devices stopped to work after `restart 1` and I had to turn off main power for few seconds multiple times to get it working and then modify `ShutterButton` properly.

I think there is missing some checking of bounds.

**Reply by stefanbode** [reply]
Yes indeed the Shutterbutton does not check the boundaries. I hoped everybody is aware. The problem of the check is that in this case you have to follow an order. First shutterconfig and then shutterbutton. I think I can avoid the awkward behavior to limit it during execution. Not on definition. So if shutterbutton set a tilt of 90 and the max is 0 it get reduced to 0. this are just two lines. Thanks for testing.

**Reply by stefanbode** [reply]
I submitted a change to fix the problem #18364

**Reply by JosefRypacek** [reply]
Thanks. Yes, this is definitely user misconfiguration, but I was scared by inaccessibility of the device anyway.

**Reply by stefanbode** [reply]
Yes this is always a problem in programming that invalid states of parameters can cause ANY unforeseen behavior. If you try to hack something this is what you are aiming for. :-)

**Comment by stefanbode** [comment]
Do you use the Shelly 2.5 to control you shutter?

**Reply by Mark-Muc** [reply]
@me?
no I use a 
```
Shelly Plus 2PM PCB v0.1.9
ESP Chip ID	9011716 (ESP32-U4WDH-D rev.3)
Flash Chip ID	0x164020 (DIO)
Größe Flash-Chip	4096 KB
Flash nutzbar	4096 KB
Größe Programm	1558 KB
Flash frei	1321 KB
Freier Arbeitsspeicher	77.3 KB (frag. 7%)
 
Partition safeboot	832 KB (used 92%)
Partition app0*	2880 KB (used 54%)
Partition fs	320 KB
```

**Reply by JosefRypacek** [reply]
Where you are missing the slider? Tasmota web UI or Home Assistant only?

Did you try to setup the shutter stuff again after upgrade?

**Reply by Mark-Muc** [reply]
WebUI 
looks like there is no sliderfunktion . only 2 buttons apeer
yes ad it said "writing to .."
i set So80 0 and again So80 1 .. but no change .. I have to damit, that I did an update on a existing 12.4.0 definde shutter!!
Mark

**Reply by JosefRypacek** [reply]
It's completely different problem then I had.

I'm using following to setup shutter (and slider is shown)

```
SetOption80 1
Interlock ON
Interlock 1,2
ShutterRelay1 1
```

Is your version similar to mine? (only build date&time should differ).

```
Program Version	[12.4.0.5](tasmota32)
Build Date & Time	2023-04-05T11:58:47
Core/SDK Version	2_0_7/[4.4.4.202]30218
```

There is also `status0` command which can be used to identify some basic settings of your device and paste here.

**Reply by Mark-Muc** [reply]
THX JosefRypacek
```
SetOption80 1
Interlock 1,2
Interlock ON
ShutterRelay1 1
```
does the trick to get the shuttterslider visible (esp. Shutterrelay1 1)
Version now (i built a solo1 for Shelly 2PM)

```
Tasmota Version	12.4.0.5 (TasmoCompiler-esp32solo1)single-core
Erstellungs-Datum & -Uhrzeit	2023.04.08 11:37:30
Core-/SDK-Version	2_0_7/4.4.4.202 30218
```
THX & Regards
Mark

**Reply by stefanbode** [reply]
Hm interesting. In esp32 shutter all env are copied into a file therefore the first start there is a migration of the old setting into the new env. Maybe there is a bug. Tested, but maybe I missed something. 

Interlock ist the same. Setoption80 also. Maybe I should check what I do with the Shutterrelay.

**Reply by stefanbode** [reply]
Can someone help me with flashing the Shelly 2.5. I just ordered 2 new ones. Do I need to press the button to be in boot mode. Normal gpio0 to ground should be enough. Additionally there is a pin described pull it high. Is this required?  Anyhow currently I do not get the device into boot mode. Normal operation still works.

**Reply by Jason2866** [reply]
@stefanbode OTA should work https://github.com/arendst/mgos-to-tasmota

**Reply by Mark-Muc** [reply]
> Hm interesting. In esp32 shutter all env are copied into a file therefore the first start there is a migration of the old setting into the new env. Maybe there is a bug. Tested, but maybe I missed something.
> 
> Interlock ist the same. Setoption80 also. Maybe I should check what I do with the Shutterrelay.

That's might be the problem by me .. pls also check for:
```
- shutterrelay1
- Shuttermode
- ShutterOpenDuration1
- ShutterCloseDuration1
- ShutterSetHalfway1
- ShutterEnableEndStopTime1
- shutterbutton1
```
This was the swiches and times I have to add after upgrading!!
Thx
Mark_Muc

I do have 3 32-Bit-Shutter capable devices I can test on!!

**Reply by stefanbode** [reply]
Yes I assume only shutterrelay1 1 is missing in the migration. The rest should be fine. Will send an update in a few minutes

**Reply by stefanbode** [reply]
Hmm the init procedure looks fine:
Start is:
```
void ShutterSettingsDefault(void) {
  // Init default values in case file is not found

  AddLog(LOG_LEVEL_INFO, PSTR("Shutter: " D_USE_DEFAULTS));

  memset(&ShutterSettings, 0x00, sizeof(ShutterSettings));
```
You should see the init message in the serial console.

Later on it loops to the shutter definition and copy values..
```
  for (uint32_t i = 0; i < MAX_SHUTTERS; i++) {
    // copy values from settings
    for (uint32_t j = 0; j < 5; j++) {
      if (j<2) ShutterSettings.shutter_pwmrange[j][i] = Settings->shutter_pwmrange[j][i];
      ShutterSettings.shutter_tilt_config[j][i] = Settings->shutter_tilt_config[j][i];
      ShutterSettings.shuttercoeff[j][i] = Settings->shuttercoeff[j][i];
    }
    ShutterSettings.shutter_tilt_pos[i] = Settings->shutter_tilt_pos[i];
    ShutterSettings.shutter_opentime[i] = Settings->shutter_opentime[i];
    ShutterSettings.shutter_closetime[i] = Settings->shutter_closetime[i];
    ShutterSettings.shutter_options[i] = Settings->shutter_options[i];
    ShutterSettings.shutter_set50percent[i] = Settings->shutter_set50percent[i];
    ShutterSettings.shutter_position[i] = Settings->shutter_position[i];
    ShutterSettings.shutter_startrelay[i] = Settings->shutter_startrelay[i];
    ShutterSettings.shutter_motordelay[i] = Settings->shutter_motordelay[i];
    
  }
```
Can someone check if shutterrelay1 to 4 is "0" if defined before and then opened in ESP32 Version. The trigger is, that there is no file found in ESP32 file system.

**Reply by Mark-Muc** [reply]
so I did:

Version: 12.3.1.6
```
10:37:19.761 CMD: backlog So80; interlock ;Shuttermode; shutterrelay1; ShutterOpenDuration1; ShutterCloseDuration1; ShutterSetHalfway1; ShutterEnableEndStopTime1; shutterbutton1
10:37:19.809 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"SetOption80":"ON"}
10:37:20.035 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Interlock":"ON","Groups":"1,2"}
10:37:20.286 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterMode":1}
10:37:20.538 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterRelay1":1}
10:37:20.786 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterOpenDuration1":29.0}
10:37:21.037 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterCloseDuration1":27.1}
10:37:21.286 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterSetHalfway1":53}
10:37:21.536 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterEnableEndStopTime1":0}
10:37:21.786 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterButton1":"-"}
```
after uppgrading to 12.4.0.5
Compilerswitches:

```
#ifdef APP_TIMEZONE
  #undef APP_TIMEZONE
#endif
#define APP_TIMEZONE 99
#ifndef USE_SHUTTER
#define USE_SHUTTER     
#endif
#define USE_SHUTTER_ESP32
#define MOTOR_STOP_TIME 750

```
do cmd:
```
10:55:55.905 CMD: backlog So80; interlock ;Shuttermode; shutterrelay1; ShutterOpenDuration1; ShutterCloseDuration1; ShutterSetHalfway1; ShutterEnableEndStopTime1; shutterbutton1
10:55:55.956 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"SetOption80":"ON"}
10:55:56.181 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Interlock":"ON","Groups":"1,2"}
10:55:56.428 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterMode":0}
10:55:56.681 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterRelay1":0}
10:55:56.931 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Command":"Error"}
10:55:57.178 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Command":"Error"}
10:55:57.429 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Command":"Error"}
10:55:57.680 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Command":"Error"}
10:55:57.930 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterButton1":"-"}

```
so it seams that non of the switches are converted!!
after:` shutterrelay1 1 `
it looks like that:

```
11:02:21.535 CMD: shutterrelay1 1
11:02:21.538 SHT: ShutterMode: 1
11:02:21.540 SHT: RESET/INIT CALIBRATION MATRIX DIV 0
11:02:21.547 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterRelay1":1}
11:02:22.101 SHUTTER: About to save settings to file /.drvset027
11:02:44.164 CMD: backlog So80; interlock ;Shuttermode; shutterrelay1; ShutterOpenDuration1; ShutterCloseDuration1; ShutterSetHalfway1; ShutterEnableEndStopTime1; shutterbutton1
11:02:44.213 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"SetOption80":"ON"}
11:02:44.439 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"Interlock":"ON","Groups":"1,2"}
11:02:44.691 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterMode":1}
11:02:44.942 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterRelay1":1}
11:02:45.191 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterOpenDuration1":10.0}
11:02:45.440 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterCloseDuration1":10.0}
11:02:45.691 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterSetHalfway1":50}
11:02:45.941 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterEnableEndStopTime1":0}
11:02:46.190 MQT: stat/Rollo-1-OG-Schlaf-Tuer/RESULT = {"ShutterButton1":"-"}

```
**Defaults are set!!**
Hope it helps to find / fix the problem
Mark_Muc

**Reply by stefanbode** [reply]
Yes , this is interesting and need further investigation. It looks like the migration procedure is not called. I hoped that at least the openduration and this survive. It did not. Therefore a general problem in my point of view. I'm later back from holiday. I need a real ESP32 to get more debug and log information. Until here: Thanks for help.

**Reply by arendst** [reply]
If you were about to upgrade to v12.4.0.5 did you make sure file ``.drvset27`` did NOT exists? Otherwise it would still use these parameters instead of the migrated ones.

**Reply by stefanbode** [reply]
yes I do test now with a clean filesystem.. give me some time.

**Reply by stefanbode** [reply]
I submitted a fix. Beginners error. I forget to save the migrated data before loading it again. Sorry.

**Reply by stefanbode** [reply]
you can easily migrate the data again by DELETING the .drvset0027 and restart after updating the firmware. The file is a HIDDEN file. you must enable this.

**Reply by stefanbode** [reply]
> @stefanbode OTA should work https://github.com/arendst/mgos-to-tasmota

hi  @Jason2866 : Yes I was able to flash one of my old Shelly 2.5 with this method. The problem with the new Shelly PLUS 2.5PM is that this device is not in the list and I did not get the mos tailored to use my own firmware. The new plus series is based on ESP32. The old one was a 2Mb Flash ESP8266EX version. Any Idea?

**Reply by stefanbode** [reply]
I found something in github. But seems not to work either:
http://github.com/tasmota/mgos32-to-tasmota32/releases/download/main/mgos32-to-tasmota32-Plus2PM.zip

The Plus2PM is correct

**Reply by stefanbode** [reply]
Ok, now I found a version that seems to work... http://ota.tasmota.com/tasmota32/shelly/mgos32-to-tasmota32-Plus2PM.zip

**Reply by stefanbode** [reply]
only seems. there was uploading and progress but at the end nothing happened..

**Reply by Jason2866** [reply]
@stefanbode tried a restart (power off on)? If Shelly firmware comes up, do the OTA flash of Tasmota again.
The switch to the new flashed OTA partition does not always work.

**Reply by stefanbode** [reply]
agree. it take some tries. now TASMOTA is on and running. THX for the support.
